client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx

```diff
@@ -0,0 +1,171 @@
+import EmailForwardingAddNewCompact from 'calypso/my-sites/email/email-forwarding/email-forwarding-add-new-compact';
+import { composeAnalytics, withAnalytics } from 'calypso/state/analytics/actions';
+import { addEmailForward } from 'calypso/state/email-forwarding/actions';
+import getEmailForwardingLimit from 'calypso/state/selectors/get-email-forwarding-limit';
+import {
+	addEmailForwardSuccess,
+	getEmailForwards,
+	isAddingEmailForward,
+} from 'calypso/state/selectors/get-email-forwards';
+import { isRequestingSiteDomains } from 'calypso/state/sites/domains/selectors';
+import { getSiteSlug } from 'calypso/state/sites/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+class EmailForwardingAddNewCompactList extends Component {
+	static propTypes = {
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
+		onConfirmEmailForwarding: PropTypes.func.isRequired,
+		onAddEmailForwardSuccess: PropTypes.func,
+		selectedDomainName: PropTypes.string.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = {
+			emailForwards: [ { destination: '', mailbox: '', isValid: false } ],
+			isRedirecting: false,
+			newEmailForwardAdded: false,
+		};
+	}
+
+	hasValidEmailForwards() {
+		return ! this.state.emailForwards?.some( ( forward ) => ! forward.isValid );
+	}
+
+	addNewEmailForwardWithAnalytics = ( domainName, mailbox, destination ) => {
+		withAnalytics(
+			composeAnalytics(
+				this.props.onConfirmEmailForwarding(),
+				this.props.addEmailForward( domainName, mailbox, destination )
+			)
+		);
+	};
+
+	addNewEmailForwardsClick = ( event ) => {
+		const { isSubmittingEmailForward, selectedSiteSlug, selectedDomainName } = this.props;
+
+		event.preventDefault();
+
+		if ( isSubmittingEmailForward ) {
+			return;
+		}
+
+		this.state.emailForwards?.map( ( forward ) => {
+			const { mailbox, destination } = forward;
+
+			this.addNewEmailForwardWithAnalytics(
+				selectedDomainName,
+				mailbox,
+				destination,
+				selectedSiteSlug
+			);
+		} );
+		this.setState( { newEmailForwardAdded: true } );
+	};
+
+	onAddNewEmailForward = () => {
+		this.setState( {
+			emailForwards: [ ...this.state.emailForwards, { destination: '', mailbox: '' } ],
+		} );
+	};
+
+	renderActionsButtons() {
+		const { translate } = this.props;
+		return (
+			<div className="email-forwarding-add-new-compact-list__actions">
+				<Button
+					primary
+					onClick={ this.addNewEmailForwardsClick }
+					disabled={ ! this.hasValidEmailForwards() }
+				>
+					{ translate( 'Add' ) }
+				</Button>
+			</div>
+		);
+	}
+
+	onRemoveEmailForward = ( index ) => {
+		const [ emailForwards ] = this.state;
+		emailForwards.splice( index, 1 );
+		this.setState( { emailForwards } );
+	};
+
+	onUpdateEmailForward = ( index, name, value ) => {
+		// eslint-disable-next-line prefer-const
+		let emailForwards = this.state.emailForwards;
+		emailForwards[ index ][ name ] = value;
```

Comment:
We're directly mutating `this.state` here, and that's forbidden. Again, as per the [docs](https://reactjs.org/docs/react-component.html#setstate):

> `state` is a reference to the component state at the time the change is being applied. It should not be directly mutated.

To fix this, you should use a copy of `this.state.emailForwards` instead of referencing the original object, for example:

```
let emailForwards = { ...this.state.emailForwards };
```